core/src/main/java/org/apache/calcite/rel/metadata/RelMetadataQuery.java

```diff
@@ -351,11 +351,32 @@ public Double getPercentageOriginalRows(RelNode rel) {
    * returns null
    */
   public RelColumnOrigin getColumnOrigin(RelNode rel, int column) {
+    return getColumnOrigin(rel, column, false);
+  }
+
+  /**
+   * Determines the origin of a column. If the param derived is false, provided the column maps
+   * to a single column that isn't derived, otherwise the column is derived.
+   *
+   * @see #getColumnOrigins(org.apache.calcite.rel.RelNode, int)
+   *
+   * @param rel the RelNode of the column
+   * @param column the offset of the column whose origin we are trying to
+   * determine
+   * @param derived allow column derived
+   *
+   * @return the origin of a column provided it's a simple column; otherwise,
+   * returns null
+   */
+  public RelColumnOrigin getColumnOrigin(RelNode rel, int column, Boolean derived) {
     final Set<RelColumnOrigin> origins = getColumnOrigins(rel, column);
     if (origins == null || origins.size() != 1) {
       return null;
     }
     final RelColumnOrigin origin = Iterables.getOnlyElement(origins);
+    if (derived) {
```

Comment:
If derived == null, we would get an NPE here.

Reply:
> If derived == null, we would get an NPE here.

Thanks, update the code.